# Healthcare Readmission Analytics

## Notebook 01 — Project Setup & Data Understanding

### Project Goal

The goal of this project is to analyze hospital readmission patterns using healthcare data from 130 US hospitals.

This notebook focuses on:

- project setup
- dataset loading
- initial data understanding
- dataset structure analysis
- data validation

### Dataset

Source:
UCI Machine Learning Repository — Diabetes 130-US Hospitals Dataset

### Why Data Understanding Matters

Understanding the dataset before cleaning and modeling helps ensure:

- correct feature selection
- reliable analysis
- proper machine learning preparation
- trustworthy business insights

#### Step 1: Import Libraries
##### 
This ensures:
- all dependencies work
- environment is configured correctly
- notebook is reproducible

Consequence if Skipped: Later cells may fail unexpectedly.


In [1]:
# ============================================
# IMPORT LIBRARIES
# ============================================

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

#### Step 2: Configure Notebook Styling

##### 
This creates professional and consistent charts throughout the project.

Consequence if Skipped: Charts may look inconsistent and less professional.

In [2]:
# ============================================
# NOTEBOOK VISUAL SETTINGS
# ============================================

sns.set_theme(
    style="whitegrid",
    palette="Set2",
    context="notebook"
)

plt.rcParams["figure.figsize"] = (10, 6)
plt.rcParams["axes.titlesize"] = 14
plt.rcParams["axes.labelsize"] = 12

#### Step 3: Load Dataset

##### 
Use of Relative Paths make projects portable.
This is critical for:
- GitHub
- deployment
- collaboration

Consequence if Skipped: Absolute paths break on other systems.

In [3]:
# ============================================
# LOAD DATASET
# ============================================

data_path = "../data/raw/diabetic_data.csv"

df = pd.read_csv(data_path)

print("Dataset loaded successfully.")

Dataset loaded successfully.


#### Step 4: Preview Dataset
#####
This helps us visually inspect:
- column names
- formatting
- categorical values
- potential issues

Consequence if Skipped: Problems may remain unnoticed until much later.

In [4]:
# ============================================
# DATA PREVIEW
# ============================================

df.head()

,encounter_id,patient_nbr,race,gender,age,weight,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,...,citoglipton,insulin,glyburide-metformin,glipizide-metformin,glimepiride-pioglitazone,metformin-rosiglitazone,metformin-pioglitazone,change,diabetesMed,readmitted
0,2278392,8222157,Caucasian,Female,[0-10),?,6,25,1,1,...,No,No,No,No,No,No,No,No,No,NO
1,149190,55629189,Caucasian,Female,[10-20),?,1,1,7,3,...,No,Up,No,No,No,No,No,Ch,Yes,>30
2,64410,86047875,AfricanAmerican,Female,[20-30),?,1,1,7,2,...,No,No,No,No,No,No,No,No,Yes,NO
3,500364,82442376,Caucasian,Male,[30-40),?,1,1,7,2,...,No,Up,No,No,No,No,No,Ch,Yes,NO
4,16680,42519267,Caucasian,Male,[40-50),?,1,1,7,1,...,No,Steady,No,No,No,No,No,Ch,Yes,NO


#### Step 5: Dataset Shape
##### 
This tells us:
- dataset size
- analysis scale
- ML feasibility

In [14]:
# ============================================
# DATASET SHAPE
# ============================================
df.shape

(101766, 50)

##### 
What this means: 
- 101,766 patient encounters
- 50 variables/features

This is a very good dataset size for:

- SQL analytics
- Healthcare KPI reporting
- Machine Learning
- Power BI dashboards
- Streamlit applications
- Why this is good

###### 
Many portfolio projects use only a few thousand rows. A dataset with over 100,000 records demonstrates experience with larger datasets.

In [15]:
print("Rows and Columns:", df.shape)

Rows and Columns: (101766, 50)


#### Step 6: Column Information

In [6]:
# ============================================
# COLUMN INFORMATION
# ============================================

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 101766 entries, 0 to 101765
Data columns (total 50 columns):
 #   Column                    Non-Null Count   Dtype
---  ------                    --------------   -----
 0   encounter_id              101766 non-null  int64
 1   patient_nbr               101766 non-null  int64
 2   race                      101766 non-null  str  
 3   gender                    101766 non-null  str  
 4   age                       101766 non-null  str  
 5   weight                    101766 non-null  str  
 6   admission_type_id         101766 non-null  int64
 7   discharge_disposition_id  101766 non-null  int64
 8   admission_source_id       101766 non-null  int64
 9   time_in_hospital          101766 non-null  int64
 10  payer_code                101766 non-null  str  
 11  medical_specialty         101766 non-null  str  
 12  num_lab_procedures        101766 non-null  int64
 13  num_procedures            101766 non-null  int64
 14  num_medications           10176

##### 
This dataset has : 13 numeric columns and 37 string columns
Numeric Features: 
- time_in_hospital
- num_lab_procedures
- num_medications
- number_outpatient
- number_emergency
- number_inpatient

Categorical Features:
- race
- gender
- age
- medical_specialty
- A1Cresult
- readmitted

##### df.info() Is Extremely Important as it helps identify:
- missing values
- incorrect data types
- memory usage
- categorical vs numerical columns

This is one of the most important commands in data analysis.

Consequence if Skipped: You may: 
- treat text as numeric
- train ML on wrong types
- create invalid charts

#### Step 7: Statistical Summary
##### 
This helps detect:
- outliers
- unusual values
- skewed distributions
- impossible values

In [7]:
# ============================================
# STATISTICAL SUMMARY
# ============================================

df.describe()

,encounter_id,patient_nbr,admission_type_id,discharge_disposition_id,admission_source_id,time_in_hospital,num_lab_procedures,num_procedures,num_medications,number_outpatient,number_emergency,number_inpatient,number_diagnoses
count,1.017660e+05,1.017660e+05,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000,101766.000000
mean,1.652016e+08,5.433040e+07,2.024006,3.715642,5.754437,4.395987,43.095641,1.339730,16.021844,0.369357,0.197836,0.635566,7.422607
std,1.026403e+08,3.869636e+07,1.445403,5.280166,4.064081,2.985108,19.674362,1.705807,8.127566,1.267265,0.930472,1.262863,1.933600
min,1.252200e+04,1.350000e+02,1.000000,1.000000,1.000000,1.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000
25%,8.496119e+07,2.341322e+07,1.000000,1.000000,1.000000,2.000000,31.000000,0.000000,10.000000,0.000000,0.000000,0.000000,6.000000
50%,1.523890e+08,4.550514e+07,1.000000,1.000000,7.000000,4.000000,44.000000,1.000000,15.000000,0.000000,0.000000,0.000000,8.000000
75%,2.302709e+08,8.754595e+07,3.000000,4.000000,7.000000,6.000000,57.000000,2.000000,20.000000,0.000000,0.000000,1.000000,9.000000
max,4.438672e+08,1.895026e+08,8.000000,28.000000,25.000000,14.000000,132.000000,6.000000,81.000000,42.000000,76.000000,21.000000,16.000000


#### Step 8: View All Columns
#####
Healthcare datasets often contain many columns. We must understand:
- feature names
- naming conventions
- target variable

In [8]:
# ============================================
# COLUMN NAMES
# ============================================

df.columns.tolist()

['encounter_id',
 'patient_nbr',
 'race',
 'gender',
 'age',
 'weight',
 'admission_type_id',
 'discharge_disposition_id',
 'admission_source_id',
 'time_in_hospital',
 'payer_code',
 'medical_specialty',
 'num_lab_procedures',
 'num_procedures',
 'num_medications',
 'number_outpatient',
 'number_emergency',
 'number_inpatient',
 'diag_1',
 'diag_2',
 'diag_3',
 'number_diagnoses',
 'max_glu_serum',
 'A1Cresult',
 'metformin',
 'repaglinide',
 'nateglinide',
 'chlorpropamide',
 'glimepiride',
 'acetohexamide',
 'glipizide',
 'glyburide',
 'tolbutamide',
 'pioglitazone',
 'rosiglitazone',
 'acarbose',
 'miglitol',
 'troglitazone',
 'tolazamide',
 'examide',
 'citoglipton',
 'insulin',
 'glyburide-metformin',
 'glipizide-metformin',
 'glimepiride-pioglitazone',
 'metformin-rosiglitazone',
 'metformin-pioglitazone',
 'change',
 'diabetesMed',
 'readmitted']

In [10]:
print("Number of columns:", len(df.columns))

Number of columns: 50


In [11]:
for i, col in enumerate(df.columns, start=1):
    print(f"{i}. {col}")

1. encounter_id
2. patient_nbr
3. race
4. gender
5. age
6. weight
7. admission_type_id
8. discharge_disposition_id
9. admission_source_id
10. time_in_hospital
11. payer_code
12. medical_specialty
13. num_lab_procedures
14. num_procedures
15. num_medications
16. number_outpatient
17. number_emergency
18. number_inpatient
19. diag_1
20. diag_2
21. diag_3
22. number_diagnoses
23. max_glu_serum
24. A1Cresult
25. metformin
26. repaglinide
27. nateglinide
28. chlorpropamide
29. glimepiride
30. acetohexamide
31. glipizide
32. glyburide
33. tolbutamide
34. pioglitazone
35. rosiglitazone
36. acarbose
37. miglitol
38. troglitazone
39. tolazamide
40. examide
41. citoglipton
42. insulin
43. glyburide-metformin
44. glipizide-metformin
45. glimepiride-pioglitazone
46. metformin-rosiglitazone
47. metformin-pioglitazone
48. change
49. diabetesMed
50. readmitted


#### Step 9: Target Variable Inspection
##### This is our ML prediction target. We must understand:
- class balance
- prediction categories
- business relevance

Consequence if Skipped: You may build incorrect ML models

In [9]:
# ============================================
# TARGET VARIABLE ANALYSIS
# ============================================

df["readmitted"].value_counts()

readmitted
NO     54864
>30    35545
<30    11357
Name: count, dtype: int64

| Readmission Status |  Count | Approx % |
| ------------------ | -----: | -------: |
| NO                 | 54,864 |    53.9% |
| >30                | 35,545 |    34.9% |
| <30                | 11,357 |    11.2% |

This column is our future target variable.

Healthcare Interpretation would be: 

| Value | Meaning                   |
| ----- | ------------------------- |
| NO    | No readmission            |
| >30   | Readmitted after 30 days  |
| <30   | Readmitted within 30 days |


###### 
Business Insight Already Identified

Approximately:46% of patients were readmitted 
because: 35545 + 11357 = 46902

This is a meaningful healthcare problem. Hospitals would be interested in reducing this number.

##### 
What We Learned From Notebook 01
Dataset Overview

✅ 101,766 records

✅ 50 columns

✅ Readmission target exists

✅ Mix of numerical and categorical features

✅ Suitable for healthcare analytics

✅ Suitable for machine learning

✅ Suitable for SQL and dashboarding

⚠️ Missing values are hidden as "?"

#### Initial Findings

#####
Observations:
- Dataset loaded successfully
- Dataset is suitable for healthcare analytics
- Dataset is suitable for machine learning
- Multiple categorical features will require preprocessing
- Missing values may be represented as "?" and require investigation
- Multiple healthcare features are available
- Both categorical and numerical columns exist
- Readmission column can be used for machine learning
- Further cleaning is required before analysis

Key Findings: 

Dataset Size: 
- 101,766 patient encounters
- 50 healthcare-related variables

Data Structure: 
- 13 numerical variables
- 37 categorical variables

Readmission Distribution: 
- NO: 54,864
- >30: 35,545
- <30: 11,357

Next Steps:
Next notebook will focus on:

- missing value analysis
- duplicate checks
- invalid value handling
- data cleaning
- preprocessing